04.ls/ls.rb

```diff
@@ -33,4 +33,133 @@ def split_files(files)
   divided_files
 end
 
+FILE_TYPE = {
+  'fifo' => 'p',
+  'characterSpecial' => 'c',
+  'directory' => 'd',
+  'blockSpecial' => 'b',
+  'file' => '-',
+  'link' => 'l',
+  'socket' => 's'
+}.freeze
+
+PERMISSION = {
+  '0' => '---',
+  '1' => '--x',
+  '2' => '-w-',
+  '3' => '-wx',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
+def show_long_format(path, file_names)
+  path_to_stat = file_names.to_h { |file_name| [File.join(path, file_name), File.lstat(File.join(path, file_name))] }
+
+  puts "total #{path_to_stat.values.sum(&:blocks)}"
+  generate_body(path_to_stat).each do |file_show_data|
+    puts format_file(file_show_data, path_to_stat.values)
+  end
+end
+
+def generate_body(path_to_stat)
+  path_to_stat.map do |absolute_path, stat|
+    {
+      permission: to_permission(stat),
+      link_count: stat.nlink.to_s,
+      owner: Etc.getpwuid(stat.uid).name,
+      group: Etc.getgrgid(stat.gid).name,
+      byte_size: stat.size.to_s,
+      time_stamp: stat.mtime.strftime(to_timestamp(stat.mtime)),
+      file_name: to_file_name(stat, absolute_path)
+    }
+  end
+end
+
+def to_timestamp(mtime)
+  half_year = 15_552_000
+  (Time.now - mtime) >= half_year ? ' %_m %_d  %Y ' : ' %_m %_d %H:%M '
+end
+
+def to_file_name(stat, absolute_path)
+  file_name = File.basename(absolute_path)
+  stat.symlink? ? "#{file_name} -> #{File.readlink(absolute_path)}" : file_name
+end
+
+def format_file(file_show_data, stat_files)
+  initial_value = 0
+  space = ' '
+  [
+    file_show_data[:permission], space * 2,
+    file_show_data[:link_count].rjust(maxlength_link(stat_files, initial_value)), space,
+    file_show_data[:owner].ljust(maxlength_uid(stat_files, initial_value)), space * 2,
+    file_show_data[:group].ljust(maxlength_gid(stat_files, initial_value)), space * 2,
+    file_show_data[:byte_size].rjust(maxlength_byte(stat_files, initial_value)),
+    file_show_data[:time_stamp],
+    file_show_data[:file_name]
+  ].join
+end
+
+def maxlength_link(stat_files, length)
+  stat_files.each do |stat|
+    length = stat.nlink.to_s.length if length < stat.nlink.to_s.length
+  end
+  length
+end
+
+def maxlength_uid(stat_files, length)
+  stat_files.each do |stat|
+    length = Etc.getpwuid(stat.uid).name.length if length < Etc.getpwuid(stat.uid).name.length
```

Comment:
おっと、これだとさらにパフォーマンスが悪化しますね。

```suggestion
  stat_files.each do |stat|
    sleep 0.1
    length = Etc.getpwuid(stat.uid).name.length if length < Etc.getpwuid(stat.uid).name.length
```

のようにsleepを各maxlength_xxx に追加して実行してみてください。

> そもそもなぜここでsleepメソッドを使うのでしょうか？ということをお聞きしたいです。

あ、たしかにその目的を伝えてなかったですね！
これはパフォーマンスの悪いロジックを体感してもらうためです。
最近のパソコンはめちゃくちゃ速いので効率の悪いロジックを書いていても体感速度はほとんど変わりません。
しかし、業務で書くような大きなプログラムになってくると効率の悪さがじわじわとシステム全体のパフォーマンスを悪化させる可能性があります。

そこで、「もし`Etc.getpwuid(stat.uid).name.length`が0.1秒ぐらいかかる重たい処理だったら？」と仮定して、パフォーマンスの悪さを体感してもらうためにsleepを入れてもらいました。

Reply:
メソッドは前回のコミットの状態に戻し、
`show_long_format`メソッド内で呼び出し、`format_file`メソッドに渡すよう変更しました。

全ファイルから各最大文字数取得 → 全ファイル分繰り返すという頓狂なことをやらかしていました😞
```ruby
def show_long_format(path, file_names)
  section_to_maxlength = fetch_maxlength(stats_by_path.values)

    puts format_file(file_data, section_to_maxlength)
```